modules/standard/Random.chpl

```diff
@@ -301,18 +305,33 @@ module Random {
                         else compilerError('choice() size type must be integral or tuple of ranges');
 
       // Return N samples
-      var samples: [1..numElements] A.eltType;
+      var samples: [1..numElements] int;
 
       if replace {
         for sample in samples {
-          var randIdx = stream.getNext(resultType=int, 1, A.size);
-          sample = A[randIdx];
+          var randVal = stream.getNext(resultType=int, 0, X.size-1);
+          var randIdx = X.dim(0).orderToIndex(randVal);
+          sample = randIdx;
         }
       } else {
-        var indices: [A.domain] int = A.domain;
-        shuffle(indices);
-        for i in samples.domain {
-          samples[i] = A[indices[i]];
+        if numElements < log2(X.size) {
+          var indices: set(int);
+          var i: int = 1;
+          while i <= numElements {
+            var randVal = stream.getNext(resultType=int, 0, X.size-1);
+            if !indices.contains(randVal) {
+              var randIdx = X.dim(0).orderToIndex(randVal);
+              samples[i] = randIdx;
+              indices.add(randVal);
+              i = i + 1;
+            }
+          }
+        } else {
+          var indices: [X] int = X;
+          shuffle(indices);
+          for i in samples.domain {
+            samples[i] = (indices[(i-1)*stride+low]);
```

Comment:
Could we `orderToIndex` here instead of computing the index manually?

Reply:
Yes, thanks for pointing out I missed that.